frame/multisig/src/lib.rs

```
@@ -524,6 +519,163 @@ impl<T: Trait> Module<T> {
+				ensure!(T::IsCallable::filter(&call), Error::<T>::Uncallable);
+
+				let result = call.dispatch(frame_system::RawOrigin::Signed(id.clone()).into());
+				let _ = T::Currency::unreserve(&m.depositor, m.deposit);
+				<Multisigs<T>>::remove(&id, call_hash);
+				Self::clear_call(&call_hash);
+				Self::deposit_event(RawEvent::MultisigExecuted(
+					who, timepoint, id, call_hash, result.map(|_| ()).map_err(|e| e.error)
+				));
+				Ok(None.into())
+			} else {
+				// We cannot dispatch the call now; either it isn't available, or it is, but we
+				// don't have threshold approvals even with our signature.
+
+				// Store the call if desired.
+				let stored = maybe_call
+					.filter(|_| store)
+					.map_or(false, |data| Self::store_call(who.clone(), &call_hash, data));
+
+				if let Some(pos) = maybe_pos {
+					// Record approval.
+					m.approvals.insert(pos, who.clone());
+					<Multisigs<T>>::insert(&id, call_hash, m);
+					Self::deposit_event(RawEvent::MultisigApproval(who, timepoint, id, call_hash));
+				} else {
+					// If we already approved and didn't store the Call, then this was useless and
+					// we report an error.
+					ensure!(stored, Error::<T>::AlreadyApproved);
+				}
+
+				// Call is not made, so the actual weight does not include call
+				Ok(Some(weight_of::as_multi::<T>(
+					other_signatories_len,
+					call_len,
+					0,
+					stored, // Call stored?
+					false, // No refund
+				)).into())
+			}
+		} else {
+			// Not yet started; there should be no timepoint given.
+			ensure!(maybe_timepoint.is_none(), Error::<T>::UnexpectedTimepoint);
+
+			// Just start the operation by recording it in storage.
+			let deposit = T::DepositBase::get() + T::DepositFactor::get() * threshold.into();
+			T::Currency::reserve(&who, deposit)?;
+
+			// Store the call if desired.
+			let stored = maybe_call
+				.filter(|_| store)
+				.map_or(false, |data| Self::store_call(who.clone(), &call_hash, data));
```

Not writing to storage if there aren't enough funds for `store_call` needs a bit of juggling here, I guess.

---

No, it's outright impossible. There's already a `reserve?` in this branch and it's impossible to do two separate `reserve?`s in a dispatch.

---

It seems to me like it's a problem with the dispatchable being too complex. If we did the storing and approval separately, then this would not be an issue.
I think it's confusing if the user passes `store == true` and then it silently fails to do so without sign of anything going wrong.

---

> No, it's outright impossible. There's already a `reserve?` in this branch and it's impossible to do two separate `reserve?`s in a dispatch.

Having a second `reserve?` is not possible, but it would be possible to check the complete amount to be reserved and error out before the first `reserve?` if it's not possible. Less clean code, but much cleaner semantics.

---

I will look into a potential multi-extrinsic refactor that should address this concern.
https://github.com/paritytech/substrate/pull/6324

It is the same concern I have to be honest. I will say though I think things are fine here. I will double check the logic, but our tests should warn us of any writes before error.

---

indeed. this is making a mountain out of a molehill.

---

:shrug: sure, your call to make

---

Just to recap: You think the case of insufficient funds is not important enough to merit an error @gavofyork ?

---

I think it was something that could have passed with a simple note in the docs. However, the code changes were not especially difficult so they're in now.